### PR TITLE
Df timeseries

### PR DESCRIPTION
**Pull request by ahilbers**
Fixes issue(s) #91 

Summary of changes in this pull request:

* Add functionality to load timeseries into `calliope.Model` from pandas dataframes. 
* Create checks that verify whether:
   - timeseries are passed when they should be (when `df=...` is specified).
   - timeseries are entered either are from CSV (`file=...`) or from dataframes (`df=...`),
      but not a combination.
   - timeseries are passed in correct format (dict of dataframes) 
* Write tests for this new software that verify whether:
    - all checks specified above work as intended
    - timeseries are entered correctly for the `national_scale` model.
* Write tests for features that already existed that verify whether:
    - timeseries are entered correctly when entered from CSV files.
* Write documentation for new functionality.

Reviewer checklist:

- [x] Test(s) added to cover contribution
- [x] Documentation updated
- [x] Changelog updated
- [ ] Coverage maintained or improved

I've also run all tests and they pass locally on my machine, with a few skips, expected passes and expected fails. I'm not sure about coverage as I haven't figured out how to test this on my machine -- let me know if it goes down too much. I've tried to write tests for all the new functionality, around 5 in total.

My commit messages on this branch have unfortunately not been very informative. If you end up using a commit message when you merge this to master, it could be something like this:

```
Add functionality to load timeseries into calliope.Model from pandas dataframes

- Write code making this possible. Timeseries to be entered in a dict of pd.DataFrame, where
each key is referenced in the YAML files via df=...
- Write checks timeseries are entered correctly.
- Write tests that the code and checks behave as required.
- Some small cosmetic changes to rest of codebase.
```

### CODE REVIEW

File: calliope/test/test_core_preprocess.py [hunk 1]
```diff
@@ -1531,3 +1620,34 @@ def test_add_max_demand_timesteps(self, model):
...
+        assert (
+            model.inputs.resource.loc['region1-1::csp'].values[8] ==
+            approx(0.263805)
+        )
+        assert (
+            model.inputs.resource.loc['region1-2::csp'].values[8] ==
+            approx(0.096755)
+        )
+        assert (
+            model.inputs.resource.loc['region1-3::csp'].values[8] ==
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/test/test_core_preprocess.py]: Happy to make this change -- I was copying the style from the tests that were already there above.

File: calliope/test/test_core_preprocess.py [hunk 2]
```diff
@@ -1531,3 +1620,34 @@ def test_add_max_demand_timesteps(self, model):
...
+        assert (
+            model.inputs.resource.loc['region2::demand_power'].values[0] ==
+            approx(-2254.098)
+        )
+        assert (
+            model.inputs.resource.loc['region1-1::csp'].values[8] ==
+            approx(0.263805)
+        )
+        assert (
+            model.inputs.resource.loc['region1-2::csp'].values[8] ==
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/test/test_core_preprocess.py]: Happy to make this change -- I was copying the style from the tests that were already there above.

File: calliope/test/test_core_preprocess.py [hunk 3]
```diff
@@ -1531,3 +1620,34 @@ def test_add_max_demand_timesteps(self, model):
...
+        assert (
+            model.inputs.resource.loc['region1::demand_power'].values[0] ==
+            approx(-25284.48)
+        )
+        assert (
+            model.inputs.resource.loc['region2::demand_power'].values[0] ==
+            approx(-2254.098)
+        )
+        assert (
+            model.inputs.resource.loc['region1-1::csp'].values[8] ==
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/test/test_core_preprocess.py]: Happy to make this change -- I was copying the style from the tests that were already there above.

File: calliope/test/test_core_preprocess.py [hunk 4]
```diff
@@ -1531,3 +1620,34 @@ def test_add_max_demand_timesteps(self, model):
...
+        from dataframes (`load_timeseries_from_dataframes`=True, called via df=...).
+        """
+
+        model = model_national
+        assert (
+            model.inputs.resource.loc['region1::demand_power'].values[0] ==
+            approx(-25284.48)
+        )
+        assert (
+            model.inputs.resource.loc['region2::demand_power'].values[0] ==
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/test/test_core_preprocess.py]: Happy to make this change -- I was copying the style from the tests that were already there above.

File: calliope/test/test_core_preprocess.py [hunk 5]
```diff
@@ -1531,3 +1620,34 @@ def test_add_max_demand_timesteps(self, model):
...
+        """
+        Timeseries data should be successfully loaded into national_scale example
+        model. This test checks whether this happens with timeseries loaded both
+        from CSV (`load_timeseries_from_dataframes`=False, called via file=...) and
+        from dataframes (`load_timeseries_from_dataframes`=True, called via df=...).
+        """
+
+        model = model_national
+        assert (
+            model.inputs.resource.loc['region1::demand_power'].values[0] ==
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/test/test_core_preprocess.py]: Happy to make this change -- I was copying the style from the tests that were already there above.

File: calliope/preprocess/model_run.py [hunk 6]
```diff
@@ -420,48 +432,82 @@ def process_timeseries_data(config_model, model_run):
...
+    dtformat = config_model.model['timeseries_dateformat']
+
+    # Generate set of all files and dataframes we want to load
+    location_config = model_run.locations.as_dict_flat()
+    model_config = config_model.model.as_dict_flat()
+
+    # Find names of csv files (file=) or dataframes (df=) called in config
+    get_names = lambda datatype, config: set([
+        v.split('=')[1].rsplit(':', 1)[0]
+        for v in config.values() if str(datatype)+'=' in str(v)
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/preprocess/model_run.py]: Again, I was copying the style present in the file already, but happy to change -- let me know.

File: calliope/preprocess/model_run.py [hunk 7]
```diff
@@ -420,48 +432,82 @@ def process_timeseries_data(config_model, model_run):
...
-            file_path = os.path.join(config_model.model.timeseries_data_path, file)
-            # load the data, without parsing the dates, to catch errors in the data
+    dtformat = config_model.model['timeseries_dateformat']
+
+    # Generate set of all files and dataframes we want to load
+    location_config = model_run.locations.as_dict_flat()
+    model_config = config_model.model.as_dict_flat()
+
+    # Find names of csv files (file=) or dataframes (df=) called in config
+    get_names = lambda datatype, config: set([
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/preprocess/model_run.py]: Again, I was copying the style present in the file already, but happy to change -- let me know.

File: calliope/preprocess/model_run.py [hunk 8]
```diff
@@ -410,7 +422,7 @@ def process_tech_groups(config_model, techs):
     return tech_groups
 
 
-def process_timeseries_data(config_model, model_run):
+def process_timeseries_data(config_model, model_run, timeseries_dataframes):
```
[the remark this answers is not in the document]
Comment (ahilbers) [reply calliope/preprocess/model_run.py]: I presume this means that too many local variables are introduced? I'm happy to split this function up if necessary.

**Comment by brynpickering** [comment]
Thanks for this @ahilbers !
We're in the process of improving the format of our code using Black (see #306). We'll get that merged into master v. soon, and it would be good to then merge that into this branch. The best way, to avoid excessive merge conflicts, is to probably run [Black](https://github.com/psf/black/) on your PR branch, then merge in our master after it has also been formatted. I'll then go through and review the PR :)

**Comment by ahilbers** [comment]
@brynpickering Thanks for getting back to me on this. Sounds good -- I agree about doing things in order as it may become a mess of merge conflicts otherwise. I'll keep on eye on #306 and update this when that one goes through.

**Comment by brynpickering** [comment]
@ahilbers latest master is ready to go now with the code formatted! You should run black on your code in this branch, then merge in latest master, then push all that back to this PR.

**Comment by ahilbers** [comment]
@brynpickering I think the pull request is ready now. I've merged the changes from master and, after some back and forth between blacken and hound, dealing with the merge conflicts and running the tests (they pass on my machine but at time of writing are running on Github), I think I've fixed the issues.

Hound still has some things to say, which you can see above. I'm more than happy to deal with these, but I left the ones I wasn't totally sure about. Let me know if you want those changes to be made too.

File: calliope/preprocess/model_run.py [hunk 9]
```diff
@@ -420,78 +454,118 @@ def process_timeseries_data(config_model, model_run):
...
 
-        _assert_timeseries_available(constraint_filenames | cluster_filenames)
+    # Generate set of all files and dataframes we want to load
+    location_config = model_run.locations.as_dict_flat()
+    model_config = config_model.model.as_dict_flat()
 
-        datetime_min = []
-        datetime_max = []
+    # Find names of csv files (file=) or dataframes (df=) called in config
+    get_names = lambda datatype, config: set(
```
[the remark this answers is not in the document]
Comment (brynpickering) [reply calliope/preprocess/model_run.py:474]: I initially resolved this, but am reverting that as we should probably be removing lambdas whenever we come to commits which edit them: reckon you could update this to a func?

**Comment by ahilbers** [comment]
@brynpickering Thanks for the comments. I've changed the last lambda to a def. The only thing hound picks up now is the `process_timeseries_data` function, which has too many `if` statements. If you want me to break that one up, I'm happy to do that.

File: calliope/preprocess/model_run.py [hunk 10]
```diff
@@ -536,39 +599,94 @@ def _parser(x, dtformat):
...
+    cluster_filenames,
+    constraint_dfnames,
+    cluster_dfnames,
+    timeseries_dataframes,
+):
+    if (
+        timeseries_dataframes is not None
+        and len(constraint_filenames | cluster_filenames) > 0
+    ):
+        raise exceptions.ModelError(
```
Comment (brynpickering) [inline calliope/preprocess/model_run.py:628]: I wonder if allowing loading either from df or file would be fine? Our process at the moment is:

find files -> load files to dictionary of DFs

Maybe we just need:

find files -> load files to dictionary of DFs -> add all other DFs that we are pointing to directly

**Comment by ahilbers** [comment]
@brynpickering Thanks for reviewing this and for the feedback. Good suggestion on the structure for reading DFs/dicts -- I'm happy to work that in. The next 2 weeks are a bit full, so it might have to be after that.

**Comment by ahilbers** [comment]
Quick update that I haven't forgotten about this -- it's been rather busy, so haven't had a chance to work these in. Still planning on doing this soon!

**Comment by brynpickering** [comment]
> Quick update that I haven't forgotten about this -- it's been rather busy, so haven't had a chance to work these in. Still planning on doing this soon!

@ahilbers just a quick bump on this - how are things looking?

File: calliope/preprocess/model_run.py (hunk 10 above)
Comment (ahilbers) [reply calliope/preprocess/model_run.py]: @brynpickering Thanks for the comment on this -- simplifed everything a bit, there wasn't really any need to restrict to either all csv or df. I've made these changes (see message below).

**Comment by ahilbers** [comment]
@brynpickering Thanks for the reminder on this. I've implemented the changes you requested (thanks for those, it makes more sense to do it that way) and added on additional test.

The tests all pass on my machine. However, on the Azure system, the issues with pyomo/enum (mentioned in #318 and #319) come up again. If the changes made in #319 pass the tests, maybe it makes sense to merge those into master, and then I'll merge those again into this pull request?

**Comment by ahilbers** [comment]
@brynpickering Further to the message above, I've merged the most recent changes from the branch in issue #319 and now all the tests pass.

**Comment by brynpickering** [comment]
Great, thanks for the updates @ahilbers. I'll take a look through it all this week.

**Pull request merged**